### PR TITLE
Enhance the HttpClientFactory article to include user feedback

### LINKED ISSUE #10146: Explain what Polly is
A link to https://github.com/App-vNext/Polly or http://www.thepollyproject.org/ at least, would be useful. A short explanation of what Polly is and how it relates to HttpClientFactory would be even better.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7b506236-b211-eb9d-b571-5efecdde76c8
* Version Independent ID: 24434b0e-0b82-e477-9440-2cc8f135df75
* Content: [Use HttpClientFactory to implement resilient HTTP requests](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#implement-your-typed-client-classes-that-use-the-injected-and-configured-httpclient)
* Content Source: [docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md](https://github.com/dotnet/docs/blob/master/docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md)
* Product: **dotnet**
* Technology: **dotnet-ebooks**
* GitHub Login: @nishanil
* Microsoft Alias: **nanil**

### LINKED ISSUE #11064: Mention explicitly that HttpClientFactory is provided through a Nuget package
The article says that the `HttpClientFactory` class is available by adding a reference to the `Microsoft.Extensions.Http` package, but it doesn't explicitly say that it's available through Nuget.

At some point, it also says:

>.NET Core 2.1 introduced a new HttpClientFactory

which might be confusing since it's not a "core" feature.

My suggestion would be to make the step of installing the library with Nuget more explicit.

Thanks.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7b506236-b211-eb9d-b571-5efecdde76c8
* Version Independent ID: 24434b0e-0b82-e477-9440-2cc8f135df75
* Content: [Use HttpClientFactory to implement resilient HTTP requests](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#implement-your-typed-client-classes-that-use-the-injected-and-configured-httpclient)
* Content Source: [docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md](https://github.com/dotnet/docs/blob/master/docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md)
* Product: **dotnet**
* Technology: **dotnet-ebooks**
* GitHub Login: @nishanil
* Microsoft Alias: **nanil**

### LINKED ISSUE #12610: Explain how to set the HttpClient's BaseAddress
In previous patterns, we would explicitly set the HttpClient's BaseAddress property. This makes the subsequent method calls a lot easier/simpler, by not having to concatenate URL string parts. You don't mention how we do that with this new factory pattern or if it's different at all.

In fact, your example shows a base address (`_remoteServiceBaseUrl`), but glosses over where/how it gets a value.

```
public class CatalogService : ICatalogService
{
    private readonly HttpClient _httpClient;
    private readonly string _remoteServiceBaseUrl;

    public CatalogService(HttpClient httpClient)
    {
        _httpClient = httpClient;
    }

    public async Task&lt;Catalog&gt; GetCatalogItems(int page, int take, 
                                               int? brand, int? type)
    {
        var uri = API.Catalog.GetAllCatalogItems(_remoteServiceBaseUrl, 
                                                 page, take, brand, type);

        var responseString = await _httpClient.GetStringAsync(uri);

        var catalog = JsonConvert.DeserializeObject&lt;Catalog&gt;(responseString);
        return catalog;
    }
}
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7b506236-b211-eb9d-b571-5efecdde76c8
* Version Independent ID: 24434b0e-0b82-e477-9440-2cc8f135df75
* Content: [Use HttpClientFactory to implement resilient HTTP requests](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#implement-your-typed-client-classes-that-use-the-injected-and-configured-httpclient)
* Content Source: [docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md](https://github.com/dotnet/docs/blob/master/docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md)
* Product: **dotnet**
* Technology: **dotnet-ebooks**
* GitHub Login: @nishanil
* Microsoft Alias: **nanil**

### LINKED ISSUE #13753: Consider retiring or reducing this page and pointing people at another
To DRY up the docs and clear down some of the outstanding open issues, perhaps just point readers to this page?

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-2.2

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7b506236-b211-eb9d-b571-5efecdde76c8
* Version Independent ID: 24434b0e-0b82-e477-9440-2cc8f135df75
* Content: [Use HttpClientFactory to implement resilient HTTP requests](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests)
* Content Source: [docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md](https://github.com/dotnet/docs/blob/master/docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md)
* Product: **dotnet**
* Technology: **dotnet-ebooks**
* GitHub Login: @nishanil
* Microsoft Alias: **nanil**

### PR DESCRIPTION
**Pull request by mvelosop**
## Summary

Closes https://github.com/dotnet/docs/issues/10146
Closes https://github.com/dotnet/docs/issues/13753
Closes https://github.com/dotnet/docs/issues/11064
Closes https://github.com/dotnet/docs/issues/12610

### CODE REVIEW

File: docs/architecture/microservices/implement-resilient-applications/implement-http-call-retries-exponential-backoff-polly.md [hunk 1]
```diff
@@ -64,16 +64,16 @@ Policy
 
 ## Additional resources
 
-- **Retry pattern**  
+- **Retry pattern** \
```
Comment (mairaw) [inline docs/architecture/microservices/implement-resilient-applications/implement-http-call-retries-exponential-backoff-polly.md]: we've had problems with loc using this formatting before. I'd keep as-is

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 2]
```diff
@@ -1,32 +1,34 @@
...
+ms.date: 08/08/2019
 ---
 # Use HttpClientFactory to implement resilient HTTP requests
 
 `HttpClientFactory` is an opinionated factory, available since .NET Core 2.1, for creating <xref:System.Net.Http.HttpClient> instances to be used in your applications.
 
 ## Issues with the original HttpClient class available in .NET Core
 
-The original and well-known [HttpClient](/dotnet/api/system.net.http.httpclient?view=netstandard-2.0) class can be easily used, but in some cases, it isn't being properly used by many developers. 
+The original and well-known [HttpClient](/dotnet/api/system.net.http.httpclient?view=netstandard-2.0) class can be easily used, but in some cases, it isn't being properly used by many developers.
```
Comment (mairaw) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
The original and well-known <xref:System.Net.Http.HttpClient> class can be easily used, but in some cases, it isn't being properly used by many developers.
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 3]
```diff
@@ -1,32 +1,34 @@
...
 
-Therefore, `HttpClient` is intended to be instantiated once and reused throughout the life of an application. Instantiating an `HttpClient` class for every request will exhaust the number of sockets available under heavy loads. That issue will result in `SocketException` errors. Possible approaches to solve that problem are based on the creation of the `HttpClient` object as singleton or static, as explained in this [Microsoft article on HttpClient usage](/dotnet/csharp/tutorials/console-webapiclient). 
+Therefore, `HttpClient` is intended to be instantiated once and reused throughout the life of an application. Instantiating an `HttpClient` class for every request will exhaust the number of sockets available under heavy loads. That issue will result in `SocketException` errors. Possible approaches to solve that problem are based on the creation of the `HttpClient` object as singleton or static, as explained in this [Microsoft article on HttpClient usage](/dotnet/csharp/tutorials/console-webapiclient).
```
Comment (mairaw) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
Therefore, `HttpClient` is intended to be instantiated once and reused throughout the life of an application. Instantiating an `HttpClient` class for every request will exhaust the number of sockets available under heavy loads. That issue will result in `SocketException` errors. Possible approaches to solve that problem are based on the creation of the `HttpClient` object as singleton or static, as explained in this [Microsoft article on HttpClient usage](../../../csharp/tutorials/console-webapiclient.md).
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 4]
```diff
@@ -1,32 +1,34 @@
...
 
 But there’s a second issue with `HttpClient` that you can have when you use it as singleton or static object. In this case, a singleton or static `HttpClient` doesn't respect DNS changes, as explained in this [issue at the .NET Core GitHub repo](https://github.com/dotnet/corefx/issues/11224).
```
Comment (mairaw) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
But there’s a second issue with `HttpClient` that you can have when you use it as singleton or static object. In this case, a singleton or static `HttpClient` doesn't respect DNS changes, as explained in this [issue](https://github.com/dotnet/corefx/issues/11224) at the dotnet/corefx GitHub repository. 
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 5]
```diff
@@ -116,7 +141,7 @@ A Typed Client is, effectively, a transient object, meaning that a new instance
 
 ### Use your Typed Client classes
 
-Finally, once you have your type classes implemented and have them registered with AddHttpClient(), you can use it anywhere you can have services injected by DI, for example in any Razor page code or any controller of an MVC web app, like in the following code from eShopOnContainers.
+Finally, once you have your typed classes implemented and have them registered with AddHttpClient(), you can use them anywhere you can have services injected by DI, for example in any Razor page code or any controller of an MVC web app, like in the following code from eShopOnContainers.
```
Comment (mairaw) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
Finally, once you have your typed classes implemented and have them registered with `AddHttpClient()`, you can use them wherever you can have services injected by DI. For example, in a Razor page code or controller of an MVC web app, like in the following code from eShopOnContainers:
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 6]
```diff
@@ -1,32 +1,34 @@
...
+- Provide a central location for naming and configuring logical `HttpClients`. For example, you may configure a client (Service Agent) that's pre-configured to access a specific microservice.
 - Codify the concept of outgoing middleware via delegating handlers in `HttpClient` and implementing Polly-based middleware to take advantage of Polly’s policies for resiliency.
-- `HttpClient` already has the concept of delegating handlers that could be linked together for outgoing HTTP requests. You register http clients into the factory plus you can use a Polly handler that allows Polly policies to be used for Retry, CircuitBreakers, etc.
-- Manage the lifetime of HttpClientMessageHandlers to avoid the mentioned problems/issues that can occur when managing `HttpClient` lifetimes yourself. 
+- `HttpClient` already has the concept of delegating handlers that could be linked together for outgoing HTTP requests. You register http clients into the factory plus you can use a Polly handler to use Polly policies for Retry, CircuitBreakers, etc.
```
Comment (mairaw) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
- `HttpClient` already has the concept of delegating handlers that could be linked together for outgoing HTTP requests. You register HTTP clients into the factory and you can use a Polly handler to use Polly policies for Retry, CircuitBreakers, and so on.
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 7]
```diff
@@ -1,32 +1,34 @@
...
+To address those mentioned issues and make the management of `HttpClient` instances easier, .NET Core 2.1 introduced a new `HttpClientFactory` that can also be used to implement resilient HTTP calls by integrating Polly with it.
+
+[Polly](http://www.thepollyproject.org/) is transient-fault-handling library that helps developers add resiliency to their applications, by using some pre-defined policies in a fluent and thread-safe manner.
 
 ## What is HttpClientFactory
 
 `HttpClientFactory` is designed to:
 
-- Provide a central location for naming and configuring logical HttpClients. For example, you may configure a client (Service Agent) that's pre-configured to access a specific microservice.
+- Provide a central location for naming and configuring logical `HttpClients`. For example, you may configure a client (Service Agent) that's pre-configured to access a specific microservice.
```
Comment (mairaw) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
- Provide a central location for naming and configuring logical `HttpClient` objects. For example, you may configure a client (Service Agent) that's pre-configured to access a specific microservice.
```

File: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md [hunk 8]
```diff
@@ -37,7 +39,7 @@ There are several ways that you can use `HttpClientFactory` in your application:
 - Use Typed Clients
 - Use Generated Clients
 
-For the sake of brevity, this guidance shows the most structured way to use `HttpClientFactory` that's to use Typed Clients (Service Agent pattern), but all options are documented and are currently listed in this [article covering HttpClientFactory usage](/aspnet/core/fundamentals/http-requests?#consumption-patterns).  
+For the sake of brevity, this guidance shows the most structured way to use `HttpClientFactory` that's to use Typed Clients (Service Agent pattern), but all options are documented and are currently listed in this [article covering HttpClientFactory usage](/aspnet/core/fundamentals/http-requests#consumption-patterns).
```
Comment (mairaw) [inline docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md]:
```suggestion
For the sake of brevity, this guidance shows the most structured way to use `HttpClientFactory`, which is to use Typed Clients (Service Agent pattern). However, all options are documented and are currently listed in this [article covering HttpClientFactory usage](/aspnet/core/fundamentals/http-requests#consumption-patterns).
```

**Review by mairaw — DISMISSED** [review]
Looks good @mvelosop. Left some comments for you to consider.

File: docs/architecture/microservices/implement-resilient-applications/implement-http-call-retries-exponential-backoff-polly.md (hunk 1 above)
Comment (mvelosop) [reply docs/architecture/microservices/implement-resilient-applications/implement-http-call-retries-exponential-backoff-polly.md]: Ok, What's better (or "less bad") using `<br />` or two (invisible) spaces?

Comment (mvelosop) [reply docs/architecture/microservices/implement-resilient-applications/implement-http-call-retries-exponential-backoff-polly.md]: I realized you're using two spaces in other PR so there goes another commit 😊

**Review by BillWagner — APPROVED** [review]
Since @mvelosop has addressed all the changes requested, I'll :shipit: now.

/cc @mairaw

**Pull request merged**